### ISSUE TITLE
List components created by sources other than odo

### ISSUE THREAD

**Issue by valaparthvi** [openshift/odo#4688]
/kind user-story

## User Story
As a user, I want to be able to see components created by other tools such as ODC(and using labels as stated by the guideline [here](https://github.com/redhat-developer/app-labels).) when I list components with odo.  So that it will be easier for me to move from a project that started on the developer console or any other tool to odo.

## Acceptance Criteria 
- [ ] It should list components when running `odo list`, considering the component's use labels mentioned [here](https://github.com/redhat-developer/app-labels). and show the source(?)
- [ ] `odo list` should be able to distinguish between local and remote components and remote components must be read-only.
- [ ] Output of `odo list` to look something like this(up for a debate) - 
      
```
➜  odo git:(listcom) ✗ odo list --all-apps
Devfile Components: 
APP      NAME      PROJECT       TYPE      STATE
app1     njs2i     myproject     njs2i     Pushed

Other Components running on the cluster(read-only): 
APP            NAME              PROJECT       TYPE
sample-app     nodejs-sample     myproject     nodejs-sample
httpd          example           myproject     example
```

## Links 
- Related Epic (mandatory): #4192 , [ODO-59](https://issues.redhat.com/browse/ODO-59)

/kind user-story

**Comment by girishramnani** [comment]
I was thinking about this and a brain dump - 

- what kind of tables should we look for to map the deployment on cluster to cluster related fields?
- Also if you find something that’s deployed using helm, it would have helm related fields? Should we leverage those? Also then wouldn’t this be in Helm territory (the helm epic)
- Should there be a separate flag in odo list to apply this feature? Else it might slow down already slow `odo list `

**Comment by valaparthvi** [comment]
>what kind of values should we look for to map the deployment on cluster to component related fields?

I am not sure I understand. Can you elaborate?


> Also if you find something that’s deployed using helm, it would have helm related fields? Should we leverage those? Also then wouldn’t this be in Helm territory (the helm epic)

For now, I think we are just going to focus on the dev console, but I guess it would be nice to keep things compatible with other tools. Also, it is given that the components created on other tools use the labels as mentioned in the [guideline](https://github.com/gorkem/app-labels/blob/master/labels-annotation-for-openshift.adoc).

> Should there be a separate flag in odo list to apply this feature? Else it might slow down already slow odo list 

For now, I would like to focus on getting them, we can fix the slowness later on.

**Comment by valaparthvi** [comment]
Scope of the issue for Sprint 201 - 
Acceptance Criteria as mentioned in the issue.
>  - [ ] It should list components when running odo list, considering the component's use labels mentioned here. and show the source(?)
> - [ ] odo list should be able to distinguish between local and remote components.

**Comment by kadel** [comment]
@valaparthvi Can you please outline how the `odo list` output would look like?
One important aspect that we need to communicate to users is that components created by other tools (like webconsole) are going to be in "read-only" mode. Users will be able to see them, but they won't be able to execute commands like `odo push` against them.

**Comment by valaparthvi** [comment]
Scope of this issue for Sprint 202 - 
tldr; Merge #4742 
- [ ] Add integration tests - `odo list`, `odo list --all-apps`, `odo list --app app`, and `odo list -o json` for non-odo components
- [ ] Add unit tests for `pkg/component/component.go:List` for devfile, s2i and non-odo components.

**Comment by valaparthvi** [comment]
@kadel How do you suggest we ensure that the components not created by `odo` are read-only?
I assume they are by default because of the select `app.kubernetes.io/managed-by: odo` and for non-odo components it's anything but that.

Do you have some other way to ensure the read only part?

**Comment by kadel** [comment]
> @kadel How do you suggest we ensure that the components not created by `odo` are read-only?
> I assume they are by default because of the select `app.kubernetes.io/managed-by: odo` and for non-odo components it's anything but that.
> 
> Do you have some other way to ensure the read only part?

I was automatically assuming that all other commands are already using  `app.kubernetes.io/managed-by: odo` selector, so we don't have to worry about this and those "external" components are already invisible for other odo commands.
Is that true?

**Comment by dharmit** [comment]
I think the components created by other tools can't be accessed beyond `odo list` anyway because we don't have their devfiles.

**Comment by valaparthvi** [comment]
>I was automatically assuming that all other commands are already using app.kubernetes.io/managed-by: odo selector, so we don't have to worry about this and those "external" components are already invisible for other odo commands.
Is that true?

True. That is my understanding.
